ledger/sandbox/src/main/scala/com/digitalasset/platform/index/ReadOnlySqlLedger.scala

```diff
@@ -9,88 +9,86 @@ import akka.actor.Cancellable
-  def owner(
+  final class Owner(
       serverRole: ServerRole,
       jdbcUrl: String,
-      ledgerId: LedgerId,
+      initialLedgerId: LedgerId,
       eventsPageSize: Int,
       metrics: Metrics,
       lfValueTranslationCache: LfValueTranslation.Cache,
-  )(implicit mat: Materializer, logCtx: LoggingContext): ResourceOwner[ReadOnlyLedger] =
-    for {
-      ledgerReadDao <- JdbcLedgerDao.readOwner(
-        serverRole,
-        jdbcUrl,
-        eventsPageSize,
-        metrics,
-        lfValueTranslationCache,
-      )
-      factory = new Factory(ledgerReadDao)
-      ledger <- ResourceOwner.forFutureCloseable(() => factory.createReadOnlySqlLedger(ledgerId))
-    } yield ledger
-
-  private class Factory(ledgerDao: LedgerReadDao)(implicit logCtx: LoggingContext) {
-
-    private val logger = ContextualizedLogger.get(this.getClass)
-
-    /**
-      * Creates a DB backed Ledger implementation.
-      *
-      * @return a compliant read-only Ledger implementation
-      */
-    def createReadOnlySqlLedger(initialLedgerId: LedgerId)(
-        implicit mat: Materializer
-    ): Future[ReadOnlySqlLedger] = {
-      implicit val ec: ExecutionContext = DEC
+  )(implicit mat: Materializer, logCtx: LoggingContext)
+      extends ResourceOwner[ReadOnlyLedger] {
+    override def acquire()(
+        implicit executionContext: ExecutionContext
+    ): Resource[ReadOnlyLedger] =
       for {
-        ledgerId <- initialize(initialLedgerId)
-        ledgerEnd <- ledgerDao.lookupLedgerEnd()
-      } yield new ReadOnlySqlLedger(ledgerId, ledgerEnd, ledgerDao)
-    }
-
-    private def initialize(initialLedgerId: LedgerId): Future[LedgerId] =
-      ledgerDao
-        .lookupLedgerId()
-        .flatMap {
-          case Some(foundLedgerId @ `initialLedgerId`) =>
-            logger.info(s"Found existing ledger with ID: $foundLedgerId")
-            Future.successful(foundLedgerId)
-          case Some(foundLedgerId) =>
-            Future.failed(
-              new LedgerIdMismatchException(foundLedgerId, initialLedgerId) with StartupException)
-          case None =>
-            Future.successful(initialLedgerId)
-        }(DEC)
+        ledgerDao <- JdbcLedgerDao
+          .readOwner(
+            serverRole,
+            jdbcUrl,
+            eventsPageSize,
+            metrics,
+            lfValueTranslationCache,
+          )
+          .acquire()
+        ledgerId <- Resource.fromFuture(initialize(ledgerDao, initialLedgerId))
+        ledgerEnd <- Resource.fromFuture(ledgerDao.lookupLedgerEnd())
+        dispatcher <- ResourceOwner
+          .forCloseable(() => Dispatcher[Offset]("sql-ledger", Offset.beforeBegin, ledgerEnd))
+          .acquire()
+        ledger <- ResourceOwner
+          .forCloseable(() => new ReadOnlySqlLedger(ledgerId, ledgerDao, dispatcher))
+          .acquire()
+      } yield ledger
   }
 
+  private def initialize(
```

Comment:
```suggestion
  private def matchOrCreateLedgerId(
```
Or something along those lines.